File: SortedSubsequenceOfSize3.py
import sys

def find3Numbers(nums):
    # Nếu số phần tử trong mảng nhỏ hơn 3, không thể tìm bộ ba
    if len(nums) < 3:
        print("No such triplet found")
        return
    
    # Khởi tạo các biến
    seq = 1  # Độ dài của dãy tăng tốt nhất hiện tại
    min_num = nums[0]  # Phần tử nhỏ nhất tính đến thời điểm hiện tại
    max_seq = -sys.maxsize - 1  # Phần tử lớn nhất của dãy tốt nhất
    store_min = min_num  # Lưu giá trị nhỏ nhất của bộ ba

    # Duyệt qua các phần tử trong mảng
    for i in range(1, len(nums)):
        if nums[i] == min_num:
            # Nếu phần tử hiện tại bằng giá trị nhỏ nhất, bỏ qua
            continue
        elif nums[i] < min_num:
            # Nếu phần tử hiện tại nhỏ hơn giá trị nhỏ nhất, cập nhật
            min_num = nums[i]
            continue
        elif nums[i] < max_seq:
            # Nếu tìm thấy phần tử lớn hơn min_num nhưng nhỏ hơn max_seq
            max_seq = nums[i]
            store_min = min_num  # Cập nhật giá trị nhỏ nhất cho dãy
        elif nums[i] > max_seq:
            # Nếu tìm thấy phần tử lớn hơn max_seq
            if seq == 1:
                store_min = min_num  # Đặt lại min_num cho dãy mới
            seq += 1  # Tăng độ dài dãy
            if seq == 3:
                # Nếu độ dài dãy đạt 3, in kết quả và kết thúc
                print(f"Triplet: {store_min}, {max_seq}, {nums[i]}")
                return
            max_seq = nums[i]
    
    # Nếu không tìm thấy bộ ba nào
    print("No such triplet found")

if __name__ == '__main__':
    # Nhập mảng từ người dùng
    nums = list(map(int, input("Enter the array elements separated by spaces: ").split()))
    
    # Gọi hàm tìm bộ ba
    find3Numbers(nums)

File: SortedSubsequenceOfSize3.c
#include <stdio.h>
#include <limits.h>
void find3Numbers(int arr[], int n) {
    if (n < 3) {
        printf("No such triplet found\n");
        return;
    }
    int smaller[n]; // Mảng lưu chỉ số phần tử nhỏ hơn ở bên trái
    int greater[n]; // Mảng lưu chỉ số phần tử lớn hơn ở bên phải
    int min_index = 0; // Chỉ số phần tử nhỏ nhất hiện tại
    smaller[0] = -1; // Không có phần tử nào bên trái phần tử đầu tiên
    // Duyệt từ trái sang phải để tìm phần tử nhỏ hơn
    for (int i = 1; i < n; i++) {
        if (arr[i] <= arr[min_index]) {
            min_index = i;
            smaller[i] = -1;
        } else {
            smaller[i] = min_index;
        }
    }
    int max_index = n - 1; // Chỉ số phần tử lớn nhất hiện tại
    greater[n - 1] = -1; // Không có phần tử nào bên phải phần tử cuối cùng
    // Duyệt từ phải sang trái để tìm phần tử lớn hơn
    for (int i = n - 2; i >= 0; i--) {
        if (arr[i] >= arr[max_index]) {
            max_index = i;
            greater[i] = -1;
        } else {
            greater[i] = max_index;
        }
    }
    // Tìm phần tử có cả phần tử nhỏ hơn ở bên trái và lớn hơn ở bên phải
    for (int i = 0; i < n; i++) {
        if (smaller[i] != -1 && greater[i] != -1) {
            printf("Triplet: %d, %d, %d\n", arr[smaller[i]], arr[i], arr[greater[i]]);
            return;
        }
    }
    printf("No such triplet found\n");
}

int main() {
    int n;
    // Nhập kích thước mảng
    printf("Enter the number of elements in the array: ");
    scanf("%d", &n);

    int arr[n];
    // Nhập các phần tử mảng
    printf("Enter the array elements: ");
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }
    find3Numbers(arr, n);
    return 0;
}
